### PR TITLE
Add zap contract for streamlined liquidity management

### PR DESCRIPTION
**Pull request by azat-hafizov**
Introduces Zap contract, a peripheral helper that improves UX by enabling users to add and remove liquidity in a single tx. The contract acts as an intermediary that orchestrates multiple steps of liquidity provision and removal, reducing txs count, gas costs and complexity for end users.

- [x] Single-tx provision: `zapIn()` allows users to LP and receive wLP token in one step
- [x] Proportional withdrawal: `zapOut()` supports unwrapping and withdrawing underlyings proportionally
- [x] Specified withdrawal: Alternative `zapOut()` flow allows specifying exact token amounts to receive
- [x] Single-asset withdrawal: `zapOutSingle()` supports withdrawing to a single asset
- [x] Security: Non-reentrant modifiers and standardized error handling
- [x] Upgradability: Uses UUPS pattern for future functionality extensions

Tests
- [x] Comprehensive unit tests for all functions
- [x] Integration tests with SPA and WLPToken contracts
- [x] Edge case testing for slippage conditions and failure modes

### CODE REVIEW

File: src/periphery/Zap.sol [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+    address public wlp;
+
+    error ZeroAmount();
+    error InsufficientAllowance();
+    error SlippageExceeded();
+    error InvalidParameters();
+    error TransferFailed();
+    error UnsupportedToken();
+
+    event ZapIn(address indexed user, uint256 wlpAmount, uint256[] inputAmounts);
```
Comment (zakrad) [inline src/periphery/Zap.sol]: its better to include receiver in event instead of constant caller which is `msg.sender`, same for `ZapOut` event

File: src/periphery/Zap.sol [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+
+    address public spa;
+    address public wlp;
+
+    error ZeroAmount();
+    error InsufficientAllowance();
+    error SlippageExceeded();
+    error InvalidParameters();
+    error TransferFailed();
+    error UnsupportedToken();
```
Comment (zakrad) [inline src/periphery/Zap.sol]: unused errors

File: src/periphery/Zap.sol [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+        if (!success) _revertBytes(data);
+
+        return abi.decode(data, (address));
+    }
+
+    /**
+     * @dev Helper function to revert with the same error message as the original call
+     */
+    function _revertBytes(bytes memory data) internal pure {
+        require(data.length > 0, TransferFailed());
```
Comment (zakrad) [inline src/periphery/Zap.sol]: transfer failed might not reflect the exact reason of failure and be confusing, we can either do high level call with interface or use a more generic error like `CallFailed()`

**Review by zakrad — CHANGES_REQUESTED** [review]

File: src/periphery/Zap.sol (hunk 1 above)
Comment (azat-hafizov) [reply src/periphery/Zap.sol]: add tracking both user and receiver

**Review by zakrad — APPROVED** [review]

File: src/periphery/Zap.sol [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+        returns (uint256 wlpAmount)
+    {
+        require(spa != address(0) && wlp != address(0), InvalidParameters());
+        require(amounts.length > 0, InvalidParameters());
+        address[] memory tokens = _getTokens(spa);
+        require(amounts.length == tokens.length, InvalidParameters());
+
+        for (uint256 i = 0; i < tokens.length; i++) {
+            if (amounts[i] > 0) {
+                IERC20(tokens[i]).safeTransferFrom(msg.sender, address(this), amounts[i]);
```
Comment (zakrad) [inline src/periphery/Zap.sol:53]: now that we accept arbitrary addresses, we better add checks before and after transfers to make sure nothing left after `transferFrom` and `_mint` at the end of function

**Review by zakrad — CHANGES_REQUESTED** [review]

File: src/periphery/Zap.sol (hunk 4 above)
Comment (azat-hafizov) [reply src/periphery/Zap.sol:53]: agreed on posting SPA address in events

**Review by zakrad — APPROVED** [review]

**Review by 0xSolDev — APPROVED** [review]
